src/intrinsic.rs

```diff
@@ -330,41 +331,24 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for EvalContext<'a, 'mir, 'tcx, super:
             "likely" | "unlikely" | "forget" => {}
 
             "init" => {
-                let size = dest_layout.size;
-                let init = |this: &mut Self, val: Value| {
-                    let zero_val = match val {
-                        Value::ByRef(ptr, _) => {
-                            // These writes have no alignment restriction anyway.
-                            this.memory.write_repeat(ptr, 0, size)?;
-                            val
-                        }
-                        // TODO(solson): Revisit this, it's fishy to check for Undef here.
-                        Value::Scalar(Scalar::Bits { defined: 0, .. }) => {
-                            match this.layout_of(dest_layout.ty)?.abi {
-                                ty::layout::Abi::Scalar(_) => Value::Scalar(Scalar::null()),
-                                _ => {
-                                    // FIXME(oli-obk): pass TyLayout to alloc_ptr instead of Ty
-                                    let ptr = this.alloc_ptr(dest_layout)?;
-                                    let ptr = Scalar::Ptr(ptr);
-                                    this.memory.write_repeat(ptr, 0, size)?;
-                                    Value::ByRef(ptr, dest_layout.align)
-                                }
+                match dest {
+                    Place::Local { frame, local } => {
+                        match self.stack()[frame].locals[local].access()? {
+                            Value::ByRef(ptr, _) => {
+                                // These writes have no alignment restriction anyway.
+                                self.memory.write_repeat(ptr, 0, dest_layout.size)?;
+                            }
+                            Value::Scalar(_) => self.write_value(ValTy { value: Value::Scalar(Scalar::null(dest_layout.size).into()), ty: dest_layout.ty }, dest)?,
+                            Value::ScalarPair(..) => {
+                                self.write_value(ValTy { value: Value::ScalarPair(Scalar::null(dest_layout.size).into(), Scalar::null(dest_layout.size).into()), ty: dest_layout.ty }, dest)?;
```

Comment:
Why does this not use the usual memory write functions, but special-case locals?

(I realize you did not change this, but I am still curious and this might be a good time to add a comment.^^)

Reply:
The memory write functions would force an allocation... at least for uninit. not sure why init is implemented the same way, it does need the allocation for bigger things.

Reply:
Okay. Would be nice to have a comment along those lines.